index.ts

```diff
@@ -53,12 +54,30 @@ export class HawkTracerClient {
             this._client = new NativeClient(option);
         }
         else {
-            this._client = new NativeClient(option.source, option.map_files);
+            this._client = new NativeClient(option.source, option.map_files, option.wait_for_server);
         }
     }
 
-    public start(): boolean {
-        return this._client.start();
+    public start(): Promise<boolean> {
+        const tryConnect = (resolve: any, reject: any) => {
+            try {
+                if (this._client.start()) {
```

Comment:
this api is really confusing. Are we trying to constantly `start()` it here? Or are we just checking and it already started but just waits for connection?

Reply:
We constantly try to start() – it's a tiny bit costly, in re-building the ClientContext on the C++. But i'm not sure we care, and what's the alternative? The hawkclient doesn't really have a capability (without a heavy refactor) to just check on a socket.

Reply:
is this the logic that was there before? how did it do it?

Reply:
I'm not sure I understand the question? Surely it's all in the diff view? In `index.ts`, the logic was very simple - just called `_client.start`

```typescript
    public start(): boolean {
        return this._client.start();
    }
```

In terms of the C++ client code, the start() method got a small refactor:
- in the past it would return true, even if the connection failed – all it cared about was if valid inputs were provided,
- now it checks the inputs but will also return false if the socket is not open.